File: speakers.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"

	tea "github.com/charmbracelet/bubbletea"
)

// wSeg is one diarised transcript segment from the whispermlx JSON output.
type wSeg struct {
	Start   float64 `json:"start"`
	End     float64 `json:"end"`
	Text    string  `json:"text"`
	Speaker string  `json:"speaker"`
}

func audioStem(audioFile string) string {
	base := filepath.Base(audioFile)
	return strings.TrimSuffix(base, filepath.Ext(base))
}

// loadSegments reads the diarised segments for a transcription output dir.
func loadSegments(outDir, audioFile string) []wSeg {
	data, err := os.ReadFile(filepath.Join(outDir, audioStem(audioFile)+".json"))
	if err != nil {
		return nil
	}
	var doc struct {
		Segments []wSeg `json:"segments"`
	}
	if err := json.Unmarshal(data, &doc); err != nil {
		return nil
	}
	return doc.Segments
}

// spkStat summarises one diarised speaker.
type spkStat struct {
	ID    string
	Name  string // user-assigned; empty until named
	Quote string // longest segment, for identification
	Dur   float64
	Share float64 // 0..1 of total talk time
}

func speakerStats(segs []wSeg) []spkStat {
	byID := map[string]*spkStat{}
	var order []string
	var total float64
	quoteLen := map[string]float64{}
	for _, s := range segs {
		id := s.Speaker
		if id == "" {
			continue
		}
		st, ok := byID[id]
		if !ok {
			st = &spkStat{ID: id}
			byID[id] = st
			order = append(order, id)
		}
		d := s.End - s.Start
		st.Dur += d
		total += d
		if d > quoteLen[id] {
			quoteLen[id] = d
			st.Quote = strings.TrimSpace(s.Text)
		}
	}
	var out []spkStat
	for _, id := range order {
		out = append(out, *byID[id])
	}
	sort.SliceStable(out, func(i, j int) bool { return out[i].Dur > out[j].Dur })
	for i := range out {
		if total > 0 {
			out[i].Share = out[i].Dur / total
		}
	}
	return out
}

// displayName resolves a speaker label: assigned name, else Speaker N in
// talk-time order.
func displayName(stats []spkStat, id string) string {
	for i, s := range stats {
		if s.ID == id {
			if s.Name != "" {
				return s.Name
			}
			return fmt.Sprintf("Speaker %d", i+1)
		}
	}
	return id
}

func fmtClock(seconds float64) string {
	s := int(seconds)
	if s >= 3600 {
		return fmt.Sprintf("%d:%02d:%02d", s/3600, (s%3600)/60, s%60)
	}
	return fmt.Sprintf("%02d:%02d", s/60, s%60)
}

// writeNamedTranscript renders the diarised segments as a readable
// markdown transcript with speaker names, merging consecutive turns and
// weaving in any markers. Returns the written path.
func writeNamedTranscript(outDir, audioFile string, segs []wSeg, stats []spkStat, markers []time.Duration) (string, error) {
	path := filepath.Join(outDir, audioStem(audioFile)+"-transcript.md")
	var b strings.Builder
	fmt.Fprintf(&b, "# Transcript — %s\n\n", audioStem(audioFile))
	fmt.Fprintf(&b, "Source: %s\n\nSpeakers: ", filepath.Base(audioFile))
	for i, s := range stats {
		if i > 0 {
			b.WriteString(", ")
		}
		fmt.Fprintf(&b, "%s (%s, %.0f%%)", displayName(stats, s.ID), fmtClock(s.Dur), s.Share*100)
	}
	b.WriteString("\n\n---\n\n")

	nextMarker := 0
	writeMarkersUpTo := func(t float64) {
		for nextMarker < len(markers) && markers[nextMarker].Seconds() <= t {
			fmt.Fprintf(&b, "**— marker %d at %s —**\n\n",
				nextMarker+1, fmtClock(markers[nextMarker].Seconds()))
			nextMarker++
		}
	}

	curSpeaker := ""
	var curStart float64
	var curText strings.Builder
	flush := func() {
		if curText.Len() == 0 {
			return
		}
		fmt.Fprintf(&b, "%s [%s]: %s\n\n",
			displayName(stats, curSpeaker), fmtClock(curStart),
			strings.TrimSpace(curText.String()))
		curText.Reset()
	}
	for _, s := range segs {
		writeMarkersUpTo(s.Start)
		if s.Speaker != curSpeaker {
			flush()
			curSpeaker = s.Speaker
			curStart = s.Start
		}
		curText.WriteString(" " + strings.TrimSpace(s.Text))
	}
	flush()
	writeMarkersUpTo(1 << 30)

	if err := os.WriteFile(path, []byte(b.String()), 0o644); err != nil {
		return "", err
	}
	return path, nil
}

// writeMarkersFile records marker timestamps next to the audio.
func writeMarkersFile(audioFile string, markers []time.Duration) string {
	if len(markers) == 0 {
		return ""
	}
	path := strings.TrimSuffix(audioFile, filepath.Ext(audioFile)) + "-markers.txt"
	var b strings.Builder
	for i, mk := range markers {
		fmt.Fprintf(&b, "marker %d  %s\n", i+1, fmtClock(mk.Seconds()))
	}
	if err := os.WriteFile(path, []byte(b.String()), 0o644); err != nil {
		return ""
	}
	return path
}

// --- post-processing hook ---

type postDoneMsg struct{ err error }

// runPostHook executes the configured post_command with context in env.
func runPostHook(cmdStr, audio, outDir, transcriptMD, markersFile string) tea.Cmd {
	return func() tea.Msg {
		c := exec.Command("sh", "-c", cmdStr)
		c.Env = append(os.Environ(),
			"SB_AUDIO="+audio,
			"SB_TRANSCRIPT_DIR="+outDir,
			"SB_TRANSCRIPT_MD="+transcriptMD,
			"SB_MARKERS="+markersFile,
		)
		out, err := c.CombinedOutput()
		if err != nil {
			tail := strings.TrimSpace(string(out))
			if len(tail) > 200 {
				tail = tail[len(tail)-200:]
			}
			return postDoneMsg{fmt.Errorf("%v: %s", err, tail)}
		}
		return postDoneMsg{nil}
	}
}

File: src/speakers.rs
use serde::Deserialize;
use std::path::{Path, PathBuf};
use std::time::Duration;

use crate::transcribe::audio_stem;

/// One diarised transcript segment from the whispermlx JSON output.
#[derive(Deserialize, Clone)]
pub struct WSeg {
    #[serde(default)]
    pub start: f64,
    #[serde(default)]
    pub end: f64,
    #[serde(default)]
    pub text: String,
    #[serde(default)]
    pub speaker: String,
}

#[derive(Deserialize)]
struct WDoc {
    #[serde(default)]
    segments: Vec<WSeg>,
}

pub fn load_segments(out_dir: &Path, audio: &Path) -> Vec<WSeg> {
    let path = out_dir.join(format!("{}.json", audio_stem(audio)));
    std::fs::read_to_string(path)
        .ok()
        .and_then(|s| serde_json::from_str::<WDoc>(&s).ok())
        .map(|d| d.segments)
        .unwrap_or_default()
}

/// One diarised speaker, ranked by talk time.
#[derive(Clone)]
pub struct SpkStat {
    pub id: String,
    pub name: String,  // user-assigned; empty until named
    pub quote: String, // longest segment, for identification
    pub dur: f64,
    pub share: f64,
}

pub fn speaker_stats(segs: &[WSeg]) -> Vec<SpkStat> {
    let mut order: Vec<String> = Vec::new();
    let mut stats: Vec<SpkStat> = Vec::new();
    let mut quote_len: std::collections::HashMap<String, f64> = Default::default();
    let mut total = 0.0;
    for s in segs {
        if s.speaker.is_empty() {
            continue;
        }
        if !order.contains(&s.speaker) {
            order.push(s.speaker.clone());
            stats.push(SpkStat { id: s.speaker.clone(), name: String::new(), quote: String::new(), dur: 0.0, share: 0.0 });
        }
        let d = (s.end - s.start).max(0.0);
        total += d;
        let st = stats.iter_mut().find(|x| x.id == s.speaker).unwrap();
        st.dur += d;
        let q = quote_len.entry(s.speaker.clone()).or_insert(0.0);
        if d > *q {
            *q = d;
            st.quote = s.text.trim().to_string();
        }
    }
    stats.sort_by(|a, b| b.dur.partial_cmp(&a.dur).unwrap_or(std::cmp::Ordering::Equal));
    for st in &mut stats {
        st.share = if total > 0.0 { st.dur / total } else { 0.0 };
    }
    stats
}

/// Assigned name, else "Speaker N" in talk-time order.
pub fn display_name(stats: &[SpkStat], id: &str) -> String {
    for (i, s) in stats.iter().enumerate() {
        if s.id == id {
            return if s.name.is_empty() { format!("Speaker {}", i + 1) } else { s.name.clone() };
        }
    }
    id.to_string()
}

pub fn fmt_clock(seconds: f64) -> String {
    let s = seconds.max(0.0) as u64;
    if s >= 3600 {
        format!("{}:{:02}:{:02}", s / 3600, (s % 3600) / 60, s % 60)
    } else {
        format!("{:02}:{:02}", s / 60, s % 60)
    }
}

/// Render the diarised segments as readable markdown with speaker names,
/// merged turns, and markers woven in. Returns the written path.
pub fn write_named_transcript(
    out_dir: &Path,
    audio: &Path,
    segs: &[WSeg],
    stats: &[SpkStat],
    markers: &[Duration],
) -> Result<PathBuf, String> {
    let stem = audio_stem(audio);
    let path = out_dir.join(format!("{stem}-transcript.md"));
    let mut b = String::new();
    b.push_str(&format!("# Transcript — {stem}\n\n"));
    b.push_str(&format!("Source: {}\n\nSpeakers: ", audio.file_name().map(|n| n.to_string_lossy().into_owned()).unwrap_or_default()));
    for (i, s) in stats.iter().enumerate() {
        if i > 0 {
            b.push_str(", ");
        }
        b.push_str(&format!("{} ({}, {:.0}%)", display_name(stats, &s.id), fmt_clock(s.dur), s.share * 100.0));
    }
    b.push_str("\n\n---\n\n");

    let mut next_marker = 0usize;
    let write_markers_up_to = |b: &mut String, t: f64, next: &mut usize| {
        while *next < markers.len() && markers[*next].as_secs_f64() <= t {
            b.push_str(&format!("**— marker {} at {} —**\n\n", *next + 1, fmt_clock(markers[*next].as_secs_f64())));
            *next += 1;
        }
    };

    let mut cur_speaker = String::new();
    let mut cur_start = 0.0;
    let mut cur_text = String::new();
    let flush = |b: &mut String, sp: &str, start: f64, text: &mut String| {
        if !text.trim().is_empty() {
            b.push_str(&format!("{} [{}]: {}\n\n", display_name(stats, sp), fmt_clock(start), text.trim()));
        }
        text.clear();
    };
    for s in segs {
        write_markers_up_to(&mut b, s.start, &mut next_marker);
        if s.speaker != cur_speaker {
            flush(&mut b, &cur_speaker, cur_start, &mut cur_text);
            cur_speaker = s.speaker.clone();
            cur_start = s.start;
        }
        cur_text.push(' ');
        cur_text.push_str(s.text.trim());
    }
    flush(&mut b, &cur_speaker, cur_start, &mut cur_text);
    write_markers_up_to(&mut b, f64::MAX, &mut next_marker);

    std::fs::write(&path, b).map_err(|e| e.to_string())?;
    Ok(path)
}

/// Record marker timestamps next to the audio; returns the path if any.
pub fn write_markers_file(audio: &Path, markers: &[Duration]) -> Option<PathBuf> {
    if markers.is_empty() {
        return None;
    }
    let path = audio.with_extension("").with_file_name(format!("{}-markers.txt", audio_stem(audio)));
    let mut b = String::new();
    for (i, m) in markers.iter().enumerate() {
        b.push_str(&format!("marker {}  {}\n", i + 1, fmt_clock(m.as_secs_f64())));
    }
    std::fs::write(&path, b).ok()?;
    Some(path)
}

/// Parse a "-markers.txt" back into durations.
pub fn load_markers_file(audio: &Path) -> Vec<Duration> {
    let path = audio.with_extension("").with_file_name(format!("{}-markers.txt", audio_stem(audio)));
    let Ok(data) = std::fs::read_to_string(path) else {
        return Vec::new();
    };
    let mut out = Vec::new();
    for line in data.lines() {
        let Some(last) = line.split_whitespace().last() else { continue };
        let mut secs = 0u64;
        let mut ok = true;
        for part in last.split(':') {
            match part.parse::<u64>() {
                Ok(v) => secs = secs * 60 + v,
                Err(_) => {
                    ok = false;
                    break;
                }
            }
        }
        if ok && last.contains(':') {
            out.push(Duration::from_secs(secs));
        }
    }
    out
}
